src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,272 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
+            ucs_assert_always(status == UCS_OK);
+        } else {
+            ret = ibv_destroy_cq(ctx->dummy_qp_ctx.cq);
+            if (ret != 0) {
+                ucs_warn("ibv_destroy_cq() returned %d: %m", ret);
+            }
+        }
+        
+        ucs_free(ctx);
+    });
+
+    kh_destroy_inplace(uct_rdmacm_cm_qp_contexts, &cm->ctxs);
+}
+
+ucs_status_t uct_rdmacm_cm_get_qp_context(uct_rdmacm_cm_t *cm,
+                                          struct ibv_context *verbs,
+                                          struct uct_rdmacm_cm_qp_context **ctx_p)
 {
-    struct ibv_cq *cq;
+    struct uct_rdmacm_cm_qp_context *ctx;
+    ucs_status_t status;
     khiter_t iter;
     int ret;
 
-    iter = kh_put(uct_rdmacm_cm_cqs, &cm->cqs,
+    iter = kh_put(uct_rdmacm_cm_qp_contexts, &cm->ctxs,
                   ibv_get_device_guid(verbs->device), &ret);
     if (ret == -1) {
-        ucs_error("cm %p: cannot allocate hash entry for CQ", cm);
-        return UCS_ERR_NO_MEMORY;
+        ucs_error("cm %p: cannot allocate hash entry for qp context", cm);
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
     }
 
     if (ret == 0) {
         /* already exists so use it */
-        cq = kh_value(&cm->cqs, iter);
+        ctx = kh_value(&cm->ctxs, iter);
     } else {
-        /* Create a dummy completion queue */
-        cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
-        if (cq == NULL) {
-            kh_del(uct_rdmacm_cm_cqs, &cm->cqs, iter);
-            ucs_error("ibv_create_cq() failed: %m");
-            return UCS_ERR_IO_ERROR;
+        /* Create a qp context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_qp_context),
+                         "qp_context");
+        if (ctx == NULL) {
+            ucs_error("cm %p: malloc qp context failed: %m", cm);
+            status = UCS_ERR_NO_MEMORY;
+            goto err_kh_del;
         }
 
-        kh_value(&cm->cqs, iter) = cq;
+        uct_rdmacm_cm_query_reserved_qpn_cap(verbs, ctx);
+        if (ctx->reserved_qpn_support) {
+            uct_rdmacm_cm_query_reserved_qpn_cap_2(verbs, ctx);
+            pthread_mutex_init(&ctx->reserved_qpn_ctx.mutex, NULL);
+            ucs_list_head_init(&ctx->reserved_qpn_ctx.blk_list);
+        } else {
+            /* Create a dummy completion queue */
+            ctx->dummy_qp_ctx.cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
+            if (ctx->dummy_qp_ctx.cq == NULL) {
+                ucs_error("ibv_create_cq() failed: %m");
+                status = UCS_ERR_IO_ERROR;
+                goto err_free_ctx;
+            }
+        }
+
+        kh_value(&cm->ctxs, iter) = ctx;
     }
 
-    *cq_p = cq;
+    *ctx_p = ctx;
     return UCS_OK;
+
+err_free_ctx:
+    ucs_free(ctx);
+err_kh_del:
+    kh_del(uct_rdmacm_cm_qp_contexts, &cm->ctxs, iter);
+out:
+    return status;
 }
 
-void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
```

Comment:
- rename to uct_rdmacm_cm_reserved_qpn_blk_add - it will not return blk_p, but just add it to the list in "ctx"
- is 'cm' used?

Reply:
@yosefe Fixed both.